libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,672 @@
+}
+
+float AP_Quicktune::limit_gain(AP_Quicktune::param_s param, float value)
+{
+    float saved_value = param_saved[uint8_t(param)];
+    if (max_reduce >= 0 && max_reduce < 100 && saved_value > 0){
+        // Check if we exceeded gain reduction
+        float reduction_pct = 100.0 * (saved_value - value) / saved_value;
+        if (reduction_pct > max_reduce){
+            float new_value = saved_value * (100 - max_reduce) * 0.01;
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Limiting %s %.3f -> %.3f", get_param_name(param), value, new_value);
+            value = new_value;
+        }
+    }
+   return value;
+}
+
+const char* AP_Quicktune::get_param_name(AP_Quicktune::param_s param)
+{
+    switch (param)
+    {
+        case param_s::RLL_P:
+            return "Roll P";
+        case param_s::RLL_I:
+            return "Roll I";
+        case param_s::RLL_D:
+            return "Roll D";
+        case param_s::PIT_P:
+            return "Pitch P";
+        case param_s::PIT_I:
+            return "Pitch I";
+        case param_s::PIT_D:
+            return "Pitch D";
+        case param_s::YAW_P:
+            return "Yaw P";
+        case param_s::YAW_I:
+            return "Yaw I";
+        case param_s::YAW_D:
+            return "Yaw D";
+        default:
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_param_name - param was %d", uint8_t(param));
+            return "UNK";
+    }
+}
+
+float AP_Quicktune::get_gain_mul()
+{
+   return expf(logf(2.0)/(UPDATE_RATE_HZ*double_time));
+}
+
+void AP_Quicktune::restore_all_params()
+{
+    for (int8_t pname = 0; pname < uint8_t(param_s::END); pname++){
+        if (BIT_IS_SET(param_changed, pname)){
+            set_param_value(param_s(pname), param_saved[pname]);
+            BIT_CLEAR(param_changed, pname);
+        }
+    }
+}
+
+void AP_Quicktune::save_all_params()
+{
+    // for pname in pairs(params) do
+    for (int8_t pname = 0; pname < uint8_t(param_s::END); pname++){
+        if (BIT_IS_SET(param_changed, pname)){
+            set_and_save_param_value(param_s(pname), get_param_value(param_s(pname)));
+            param_saved[pname] = get_param_value(param_s(pname));
+            BIT_CLEAR(param_changed, pname);
+        }
+    }
+}
+
+AP_Quicktune::param_s AP_Quicktune::get_pname(AP_Quicktune::axis_names axis, AP_Quicktune::stages stage)
+{
+    switch (axis)
+    {
+        case axis_names::RLL:
+            switch (stage)
+            {
+                case stages::P:
+                    return param_s::RLL_P;
+                case stages::D:
+                    return param_s::RLL_D;
+                case stages::FLTT:
+                    return param_s::RLL_FLTT;
+                case stages::FLTD:
+                    return param_s::RLL_FLTD;
+                default:
+                    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_pname - axis was %d, stage was %d", uint8_t(axis), uint8_t(stage));
```

Comment:
We do have internal error calls for such thing. You could throw a flow of control.